### DISCUSSION TITLE
Add support for "Creating New View" for JupyterLab

### DISCUSSION THREAD

**Discussion by pnsvk** [bokeh/bokeh#12032] [General]
### Software versions

Python Version: 3.9
JupyterLab: 3.x

### Browser name and version

Chrome. Any recent version

### Jupyter notebook / Jupyter Lab version

JupyterLab 3.x

### Expected behavior

The expected behavior is to have the new-view reflect that cell's output exactly as-is.

### Observed behavior

The outcome has a duplicated rendering of the same visual right under the already available output.

The same issue has already been raised in JupyterLab repo long time back, but no response (below link).
https://github.com/jupyterlab/jupyterlab/issues/779

Snapshots:

![AE330BE4-1F87-464E-90BB-DA1BD60FF956_1_105_c](https://user-images.githubusercontent.com/30823233/156518850-dba7c4b7-dd48-49c9-9ef3-af3c86dd5221.jpeg)
7
![1C3B4BE8-7FB1-4A64-BF3A-1503585246AF](https://user-images.githubusercontent.com/30823233/156518865-c13208ac-3cf8-4865-b03b-45f3e037230b.jpeg)
![11FE9243-48F8-4CB6-8B8D-C8434454A285_1_105_c](https://user-images.githubusercontent.com/30823233/156518869-9b12df94-e109-4d65-8278-4abfe883f1cf.jpeg)

### Example code

```Python
from bokeh.io import push_notebook, show, output_notebook
from bokeh.layouts import row
from bokeh.plotting import figure

output_notebook()

p1 = figure(plot_width=250, plot_height=250)
r1 = p1.circle([1, 2, 3], [4, 5, 6], size=20)

p2 = figure(plot_width=250, plot_height=250)
r2 = p2.circle([1, 2, 3], [4, 5, 6], size=20)

t = show(row(p1, p2), notebook_handle=True)
```

**Comment by bryevdv** [comment]
I've never heard of or used "Create new view for output" what is it meant to do? 

Also, the linked issue does not appear related?

**Reply by pnsvk** [reply]
@bryevdv sad to know such a popular feature of jupyterlab was unnoticed.. to us, its one of the killer features.

In this snapshot, moving the slider on left tab changes the loader on the right tab, which actually is just the cell output up above in the left-tab..

Since both the controls (slider and progressbar) cannot be accommodated in the same view-port in the left tab, we can stick it on the right, avoiding to scroll to top every time, the widget is interacted with.

![E77901A3-C8A6-4C16-9F73-775CFBD20C5E](https://user-images.githubusercontent.com/30823233/156697765-7f7c0ce1-d016-4c3a-b068-597a578f8277.jpeg)

**Reply by bryevdv** [reply]
@pnsvk I'm just not much of a notebook user, in fact, I generally try to avoid them. Speaking plainly, it would be great for Bokeh if some folks invested in notebook use and especially with JupyterLab dev expertise decided to become involved contributors to help maintain this corner of Bokeh.

**Reply by bryevdv** [reply]
That said, your description and image above seem a bit at odds with the earlier description and linked HV issue. The "cell output up above in the left-tab" (for *In[8]*) seems to no longer be visible/present at all in your image. Is "Create new view for output" supposed to *clone* the output (leaving the original in place) or *move* the output? This is a critical important distinction, as the latter is likely to be much more feasible in terms of interaction with `push_notebook`, given the way Bokeh documents and notebooks interact at a low level. (cc @philippjfr do you concur?)

**Reply by philippjfr** [reply]
As far as I know it clones the output which is the main issue because bokeh does not like two models with the same IDs on the same page. For the interactivity part I'd have to look at the internals, but I think the view should still be able to access the comm that jupyter_bokeh exposes.

**Reply by bryevdv** [reply]
FWIW I think moving is actually what would be useful for users 🤷 i.e. there is a plot created some cells above and then several cells later some ipywidgets to update the plot with `push_notebook` but oops! you can't see both on the page at once in the vertical layout. So moving the plot output to the side would be great. But (agreeing) it would need to be a *move*, in order for the output to stay connected to any existing notebook comms handle. Alternatively, if "cloning" is mandatory, I guess the comms handle could be moved to the new copy, leaving the original plot output "dead".  I think that is not-great, but more useful than nothing, I guess.

**Reply by pnsvk** [reply]
can you please watch this clip ?
just a few seconds of play from it shows the expected behavior.

https://youtu.be/1ndo6C1KWjI?t=484

**Reply by bryevdv** [reply]
Well, that's unfortunate, since I doubt keeping both views present and in-sync is doable on any kind of near-term time frame, except possibly in the one case of a Bokeh server app, since it would be the Bokeh server that does the synchronization (that's already its primary job). My [speculation below](https://github.com/bokeh/bokeh/discussions/12032#discussioncomment-2289993) about what seems possible/not possible remains unchanged.

**Reply by pnsvk** [reply]
Ok.. if not "in sync", can we have it properly rendered in the "New View" ? Which I see is the case with libraries like Altair.
Here, brush-interaction with graph in right-tab has no effect on the graph rendered in left.

![33AADF84-19BA-45CD-8892-ADA92004F45E_1_105_c](https://user-images.githubusercontent.com/30823233/157143022-3b98f9ec-c973-423a-b7b3-a8957c1c6a54.jpeg)

**Reply by bryevdv** [reply]
I’ve said as much, a few times: 

> Even in the standalone case, e.g. if you pan one plot, I would not be expect the other plot to pan along as well. They would be independent copies.

> Alternatively, if "cloning" is mandatory, I guess the comms handle could be moved to the new copy, leaving the original plot output "dead". I think that is not-great, but more useful than nothing, I guess.

That said, it is of course predicated on someone deciding to tackle the issue.

**Comment by philippjfr** [comment]
Related issue and discussion https://github.com/holoviz/pyviz_comms/issues/3

**Comment by bryevdv** [comment]
@philippjfr i see. Very off the cuff speculation:

First, I assume there is some hook in the process that we can make use of to customize or control the behavior? If not, options will be very limited in general. But assuming there is:

* Standalone content w/o notebook comms is almost certainly doable. It will be simpler if there is a hook that could call Python code, since then it could just do whatever `show` currently does to be able to make `show(p)` workable in two successive cells. But even a JS hook could probably make things work.

* Bokeh server content also seems workable, at least in the sense that the new output area should be able to just open a new session easily enough. But the two views would not be linked. Maybe you could re-open the existing session if there is a python hook to call `server_session` appropriately. Again, it really depends how much control we can insert. 

* I don't know how to make `push_notebook` work. There is only one notebook handle, it's only returned to Python, and it can only be used to update one output cell. `push_notebook` was always a bit of hack to begin with, that only existed before the Bokeh server was viable. So if Bokeh server sessions van be made to work, I think that is the priority and we just declare `push_notebook` explicitly incompatible with this JLab feature. 


But I should be clear: having the different outputs be "linked" is probably not feasible. Even in the standalone case, e.g. if you pan one plot, I would not be expect the other plot to pan along as well. They would be independent *copies*. (That's actually the number 1 reason I can't picture `push_notebook` being workable). The only circumstance where complete synchronization might be feasible is in the  Bokeh server app case, if existing sessions can be embedded (instead of creating a new session).

**Comment by bryevdv** [comment]
Given that there is so much uncertainty and unknowns around both the requirements and what is possible from the JLab side, I am converting this to a development discussion
